### ISSUE TITLE
How can I pass importObject to a WebAssembly module with Webpack 5 as a building tool?

### ISSUE THREAD

**Issue by Becavalier** [webpack/webpack#11893]
## Feature request

**What is the expected behavior?**

There should be a way to pass importObject for the instantiation of some special WebAssembly modules. So, how can I do that with Webpack 5 as a building tool?

Let's say, for example, to use the below Wasm module, I need a way to pass the implementation of the function "minus" to instantiate the module instance.

```
import { cal } from './program.wasm'

console.log(cal(5, 2)) // -3.
```

Currently, I get this error while running the app:

```
program.wasm:1 Uncaught (in promise) Error: Cannot find module 'env'
    at webpackMissingModule (program.wasm:1)
    at eval (program.wasm:1)
    at Object../src/program.wasm (src_program_wasm.bundle.js:23)
    at __webpack_require__ (main.bundle.js:494)
    at Function.fn (main.bundle.js:687)
```

**What is motivation or use case for adding/changing the behavior?**

**How should this be implemented in your opinion?**

**Are you willing to work on this yourself?**
yes

**Comment by evilebottnawi** [comment]
Thank you for creating this issue. However, issues need to follow one of our templates so that we can clearly understand your particular circumstances.

Please help us help you by recreating the issue using one of our templates.

Do not ignore the issue template questions

**Comment by congineer** [comment]
Hi there. The ability to instantiate wasm with importObject is paramount for 99.99999(9)% of real world use cases. 
I'll elaborate the ticket to re-open it:

**What is motivation or use case for adding/changing the behavior?**

Only trivial wasm examples like "fibonacci"  have 'export' fields, but  no 'import' dependencies to interoperate with JS.
Webpack 5 wasm import with config `experiments.asyncWebAssembly: true `and `module.rules: { test: /\.wasm$/, type: 'webassembly/async' }` works perfectly fine for those trivial cases of wasm containing 'export' fields only. 

Most bigger examples generated with Cheerp, Emscripten, etc. have wasm containing 'export' AND 'import' fields (for JS dependencies referred by some as 'glue code').

At present, Webpack 5 wasm import results in a WebAssemply Instance with, what seems to me, an empty  'importObj'.
Hence, trivial wasm instantiates fine, but any wasm with 'import' fields throws 'LinkError: WebAssembly.instantiate():' as it should.

Searching webpack repo for 'importObj' yields 4 code results indicating presence of [const  importObj in lib/wasm-async/AsyncWebAssemblyJavascriptGenerator.js on line 142](https://github.com/webpack/webpack/blob/827c52eeb798df4ec0733c1827d819e70998f68c/lib/wasm-async/AsyncWebAssemblyJavascriptGenerator.js#L142)

**How should this be implemented in your opinion?**

a) provide option for passing  importObject upon import, e.g. 

`const testPromise = import('./test.wasm')(importObj)`

OR

b)  provide 'experiments.asyncWebAssembly**AsModule**', i.e.  the ability to import wasm as a wasm-module instead of wasm-instance. At present [lib/wasm-async/AsyncWasmChunkLoadingRuntimeModule.js] (https://github.com/webpack/webpack/blob/af52c8f0457b1055098d6b7b545b257404d06c93/lib/wasm-async/AsyncWasmChunkLoadingRuntimeModule.js) goes too far and instantiates  
with WebAssembly.instantiateStreaming() and fallback WebAssembly.instantiate().

Having wasm imported as wasm-module instead of wasm-instance will allow passing importObj and instantiate manually according to the spec: [WebAssembly/Module](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WebAssembly/Module)

**Comment by HarikrishnanBalagopal** [comment]
Why is this issue closed?

**Comment by congineer** [comment]
> Why is this issue closed?

@[alexander-akait](https://github.com/alexander-akait) ignored my elaborated comment and did not re-open the issue. However, I found a complex way to work around this issue using type: 'asset/resource' instead of 'webassembly/async' to load the wasm file and monkey-patch it to inject the import object.

**Comment by HarikrishnanBalagopal** [comment]
@alexander-akait this is still an issue. If it is possible to pass a `importObject` to wasm somehow then please point me to the relevant documentation. The only thing I found is `asyncWebAssembly` in https://webpack.js.org/configuration/experiments/

**Comment by HarikrishnanBalagopal** [comment]
> > Why is this issue closed?
> 
> @[alexander-akait](https://github.com/alexander-akait) ignored my elaborated comment and did not re-open the issue. However, I found a complex way to work around this issue using type: 'asset/resource' instead of 'webassembly/async' to load the wasm file and monkey-patch it to inject the import object.

can you elaborate on the workaround?

**Comment by congineer** [comment]
> > > Why is this issue closed?
> > 
> > 
> > @[alexander-akait](https://github.com/alexander-akait) ignored my elaborated comment and did not re-open the issue. However, I found a complex way to work around this issue using type: 'asset/resource' instead of 'webassembly/async' to load the wasm file and monkey-patch it to inject the import object.
> 
> can you elaborate on the workaround?

@[HarikrishnanBalagopal](https://github.com/HarikrishnanBalagopal) please check in documentation if this issue is solved already. If not, please reply here and I'll do a demo with a screencast to show my hack.  Also, I'll use it to open a new issue for @[alexander-akait](https://github.com/alexander-akait) here. I'm quite busy right now, but will reserve time for it in 2-3 weeks if you really need a solution for that problem.

**Comment by alexander-akait** [comment]
@congineer Can you provide reproducible example? Nothing to use `env`, so maybe problem in your env, need investigate...

**Comment by congineer** [comment]
> 

[lexander-akait](https://github.com/alexander-akait) what does env has to do with it? In my comment  Apr 13, 2021 I clearly explained that webpack code has no means to handle the import object: "importObject is not being passed along with fetched binary". If you confirm that you are willing to investigate the issue I'll schedule to create an elaborated demo within 2-3 weeks in my agenda.

**Comment by congineer** [comment]
@HarikrishnanBalagopal and @alexander-akait what is the status of your investigations? Has this issue been resolved? Will this ticket be re-opened?

**Comment by alexander-akait** [comment]
@congineer Can you create reproducible example?

**Comment by congineer** [comment]
> @congineer Can you create reproducible example?

@alexander-akait Ok, I will write a test case and let you know here.

**Comment by braulio-balanza** [comment]
Hi @congineer, do you know if this issue has been solved? Seems like a no-brainer feature...

**Comment by congineer** [comment]
Hi @braulio-balanza. My project's WASM part shifted from spring towards December. Hence, I'll report here with examples after I'll dig into it. I have a workaround for Cheerp compiler, but I assume that the community is more interested in Emscripten example.

**Comment by braulio-balanza** [comment]
Hi @congineer, thank you for your reply. Actually, Go compiled wasm requires the use of an import object defined in the glue javascript code. Without this import object, you cannot bundle even the simplest hello world program, making webpack unusable with wasm code compiled from Go. If necessary, I could pull up a hello world example real quick that shows the need for an import object.

**Comment by congineer** [comment]
Yes, @braulio-balanza, such example could potentially help open this ticket.

**Comment by larabr** [comment]
I bumped into this issue as well, as I'd need to pass `WebAssembly.Memory` through the `importObjects`. 

I have resorted to using the [wasm-loader](https://github.com/ballercat/wasm-loader) plugin . It's marked as deprecated (due to webpack's official support of wasm) but it works well for me.

Since an example was requested, I am going to copy-paste one from the README of wasm-loader:
```js
import makeFactorial from 'wasm/factorial';

makeFactorial({
  'global': {},
  'env': {
    'memory': new WebAssembly.Memory({initial: 100, limit: 1000}),
    'table': new WebAssembly.Table({initial: 0, element: 'anyfunc'})
  }
}).then(instance => { /* code here */ });
```
But more in general, details about the `importObjects` that `WebAssembly.instantiate` accepts can be found in: https://developer.mozilla.org/en-US/docs/WebAssembly/JavaScript_interface/instantiate#syntax .

As the others have said, it's quite fundamental to be able to pass this input.

**Comment by petercockerell-brt** [comment]
Any update on this? I just came across it today. The code Webpack generates for loading the WASM has an error built in to the call to the loader:

```
/***/ "./src/wasm-code.wasm":
/*!****************************!*\
  !*** ./src/wasm-code.wasm ***!
  \****************************/
/***/ ((module, exports, __webpack_require__) => {

"use strict";
Object(function webpackMissingModule() { var e = new Error("Cannot find module 'env'"); e.code = 'MODULE_NOT_FOUND'; throw e; }());
module.exports = __webpack_require__.v(exports, module.id, "9e16aacb11ad6b85b0d4", {
	"env": {
		"memory": Object(function webpackMissingModule() { var e = new Error("Cannot find module 'env'"); e.code = 'MODULE_NOT_FOUND'; throw e; }())
	}
});

```

Without a way of setting the "env" in the module doing the import of the wasm code, it seems impossible to use this feature (where the WASM doesn't provides its own memory and imports).

**Comment by kdonahue-su** [comment]
I am also running into this issue trying to use Emscripten to generate an ES6 module.  I was unsuccessful with `wasm-loader`.  Has anyone found a workaround?

**Comment by alexander-akait** [comment]
Let's close, we are supporting `import source library from './library.wasm';` using `experiments.sourceImport: true`, so you can use this code:

```js
// you can also use comment /* webpackSource: true */
// import library from /* webpackSource: true */ './library.wasm';
import source library from './library.wasm';

const instance1 = await WebAssembly.instantiate(libraryModule, importObject1);
```
